### DISCUSSION TITLE
Memory Leak with Pluto Hooks - leading to termination of workers

### DISCUSSION THREAD

**Discussion by schlichtanders** [fonsp/Pluto.jl#2587] [Questions about using Pluto]
Hi there,

I am using PlutoHooks and occasionally running into `Distributed.ProcessExitedException(3)` with a request to restart the notebook.

Looking at the output of Pluto shell I get
```julia
Worker 3 terminated.
UNHANDLED TASK ERROR: EOFError: read end of file
Stacktrace:
 [1] (::Base.var"#wait_locked#715")(s::Sockets.TCPSocket, buf::IOBuffer, nb::Int64)
   @ Base ./stream.jl:947
 [2] unsafe_read(s::Sockets.TCPSocket, p::Ptr{UInt8}, nb::UInt64)
   @ Base ./stream.jl:955
 [3] unsafe_read
   @ ./io.jl:761 [inlined]
 [4] unsafe_read(s::Sockets.TCPSocket, p::Base.RefValue{NTuple{4, Int64}}, n::Int64)
   @ Base ./io.jl:760
 [5] read!
   @ ./io.jl:762 [inlined]
 [6] deserialize_hdr_raw
   @ /usr/local/julia/share/julia/stdlib/v1.9/Distributed/src/messages.jl:167 [inlined]
 [7] message_handler_loop(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
   @ Distributed /usr/local/julia/share/julia/stdlib/v1.9/Distributed/src/process_messages.jl:172
 [8] process_tcp_streams(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
   @ Distributed /usr/local/julia/share/julia/stdlib/v1.9/Distributed/src/process_messages.jl:133
 [9] (::Distributed.var"#103#104"{Sockets.TCPSocket, Sockets.TCPSocket, Bool})()
   @ Distributed ./task.jl:514
```

I got this in very different contexts, but only when using PlutoHooks so far.
No recording is yet ready, neither a notebook which can reproduce this (it seems random sofar, not sure what is causing it yet). Hence no issue, but a discussion.

Does someone have a clue what is going on?
Any help or hint is highly appreciated.

**Comment by schlichtanders** [comment]
one time I also got an accompanying error
```julia
Worker 4 terminated.
UNHANDLED TASK ERROR: IOError: read: connection reset by peer (ECONNRESET)
Stacktrace:
  [1] wait_readnb(x::Sockets.TCPSocket, nb::Int64)
    @ Base ./stream.jl:410
  [2] (::Base.var"#wait_locked#715")(s::Sockets.TCPSocket, buf::IOBuffer, nb::Int64)
    @ Base ./stream.jl:949
  [3] unsafe_read(s::Sockets.TCPSocket, p::Ptr{UInt8}, nb::UInt64)
    @ Base ./stream.jl:955
  [4] unsafe_read
    @ ./io.jl:761 [inlined]
  [5] unsafe_read(s::Sockets.TCPSocket, p::Base.RefValue{NTuple{4, Int64}}, n::Int64)
    @ Base ./io.jl:760
  [6] read!
    @ ./io.jl:762 [inlined]
  [7] deserialize_hdr_raw
    @ /usr/local/julia/share/julia/stdlib/v1.9/Distributed/src/messages.jl:167 [inlined]
  [8] message_handler_loop(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
    @ Distributed /usr/local/julia/share/julia/stdlib/v1.9/Distributed/src/process_messages.jl:172
  [9] process_tcp_streams(r_stream::Sockets.TCPSocket, w_stream::Sockets.TCPSocket, incoming::Bool)
    @ Distributed /usr/local/julia/share/julia/stdlib/v1.9/Distributed/src/process_messages.jl:133
 [10] (::Distributed.var"#103#104"{Sockets.TCPSocket, Sockets.TCPSocket, Bool})()
    @ Distributed ./task.jl:514
```

**Reply by schlichtanders** [reply]
@fonsp do you have a clue? It might also be related to some changes I did to Pluto (I don't think so actually, but not sure). The EOFError is the typical one which crashes my notebooks like 3 times per day.

**Reply by schlichtanders** [reply]
I found a related discourse discussion which reports exactly the same two errors.
https://discourse.julialang.org/t/problem-with-using-distributed-for-a-monte-carlo-simulation/95508

The best guess there is that it has something to do with out of memory.

This indeed may be the case - I have a long running Pluto job which uses Hooks to get regular updates (per second). If something is not cleaned up it may stack up

**Reply by schlichtanders** [reply]
I short test with `meminfo_procfs()` from https://discourse.julialang.org/t/how-to-track-total-memory-usage-of-julia-process-over-time/91167/6?u=schlichtanders indeed shows that the memory is constantly increasing.

Some memory leak. 😬

**Reply by schlichtanders** [reply]
here an example macro of mine, which repeatedly (e.g. once per second), reads from a channel and retriggers Pluto

```julia
"""
	nextvalue = @repeat_take! channel

This will repeatedly fetch for the next element from the given channel, re-evaluating
the cell each time a new value arrives.
"""
macro repeat_take!(channel)
    PlutoHooks.is_running_in_pluto_process() || return quote
		take!($(esc(channel)))
	end

	taskref = Ref{Task}()
	quote
		# kill a possibly running task.
		# even if this is refreshed by hook, there should not be any task running, hence
		# this is almost a no-op.
		# Note that this really needs to be run before init, as init may take! and block
		# so that a possibly still running but invalid task would interfere.
		$(create_taskref_cleanup(taskref))()

		# use_did_deps_change([]) will return true if the cell id changes
		# or the given variables. Normally the cell-id only changes when the
		# cell is rerun manually. Exactly that often we need to register the
		# cleanup function.
		if PlutoHooks.@use_did_deps_change([])
			register_cleanup_fn = PlutoHooks.@give_me_register_cleanup_function()
			register_cleanup_fn($(create_taskref_cleanup(taskref)))
		end

		channel = $(esc(channel))
		update, set_update, hooked = @use_state_reinit(take!(channel))
		isa(update, ExceptionFromTask) && throw(update)

		taskref = $taskref
		taskref[] = Task() do
			_channel = channel
			set_update(take!(_channel))
		end
		schedule(taskref[])
		# because every exception in the Channel is stable
		# we can do the Exception check after scheduling the task
		# this way adapting other cells can indeed auto-update this cell
		errormonitor_pluto(set_update, taskref[])
		update
	end
end
```

Any help is highly appreciated about what could cause the memory leak

**Reply by Pangoraw** [reply]
How much does the memory increase per cell run? It may not necessarely be a memory leak per say but rather that it is allocating permanent objects (for the unnamed closure in `Task` for example) on each run.

**Reply by schlichtanders** [reply]
like 1 MB in 6 runs

(I now also included GC.gc() up front - was forgetting it, the increase is less, but still there and steadily summing up.

**Reply by schlichtanders** [reply]
using `varinfo()` as suggested in Zulip it is seen that old modules are not cleaned up. Hence even if I would reduce the memory footprint of the macro, it is only a bit later that I run into OOM.

It would really be necessary to either properly garbage collect these old modules. Or to reuse the workspace without creating new ones every time.

**Reply by schlichtanders** [reply]
@Pangoraw I found that it is indeed because of a task, but haven't found any help about how to investigate this or fix this.
Do you have some tips what to do about Tasks stacking up?

**Reply by schlichtanders** [reply]
it turns out I run into a new memory leak from Julia Base. opened an issue there https://github.com/JuliaLang/julia/issues/50345

**Comment by schlichtanders** [comment]
it turns out I run into a new memory leak from Julia Base. opened an issue there https://github.com/JuliaLang/julia/issues/50345
